### PR TITLE
feat: store generated authentication secrets in secrets manager on deploy

### PR DESCRIPTION
**Pull request by Aenimus**
#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: pkg/apihandler/api.go [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
 	"go.uber.org/zap/zapcore"
 
 	"github.com/wundergraph/wundergraph/pkg/wgpb"
 )
 
+const (
+	WgEnvCsrfSecret       = "WUNDERGRAPH_CSRF_TOKEN_SECRET"
+	WgEnvHashKey          = "WUNDERGRAPH_SECURE_COOKIE_HASH_KEY"
+	WgEnvBlockKey         = "WUNDERGRAPH_SECURE_COOKIE_BLOCK_KEY"
+	validSecretCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
```
Comment (Aenimus) [inline pkg/apihandler/api.go:18]: Are other characters valid?

File: pkg/apihandler/api.go [hunk 2]
```diff
@@ -41,6 +113,7 @@ type Api struct {
 	S3UploadConfiguration []*wgpb.S3UploadConfiguration
 	Webhooks              []*wgpb.WebhookConfiguration
 	Options               *Options
+	CookieBasedSecrets    *CookieBasedSecrets
```
Comment (Aenimus) [inline pkg/apihandler/api.go:116]: Should I move CookieBasedSecrets into AuthenticationConfig > CookieBased? Or leave it as a top level field?

File: packages/sdk/src/configure/index.ts [hunk 3]
```diff
@@ -253,10 +253,13 @@ export interface ResolvedWunderGraphConfig {
 			revalidateAuthentication: boolean;
 			postLogout: boolean;
 		};
-		cookieSecurity: {
-			secureCookieHashKey: ConfigurationVariable;
-			secureCookieBlockKey: ConfigurationVariable;
-			csrfTokenSecret: ConfigurationVariable;
+		cookieSecurity?: {
+			/** @deprecated: Not used anymore */
```
Comment (jensneuse) [inline packages/sdk/src/configure/index.ts:280]: Add links to docs for replacement/alternative?

File: packages/sdk/src/configure/index.ts [hunk 4]
```diff
@@ -253,10 +253,13 @@ export interface ResolvedWunderGraphConfig {
...
+			secureCookieHashKey?: any;
+			/** @deprecated: Not used anymore */
```
Comment (jensneuse) [inline packages/sdk/src/configure/index.ts:282]: same

File: packages/sdk/src/configure/index.ts [hunk 5]
```diff
@@ -253,10 +253,13 @@ export interface ResolvedWunderGraphConfig {
...
+			secureCookieBlockKey?: any;
+			/** @deprecated: Not used anymore */
```
Comment (jensneuse) [inline packages/sdk/src/configure/index.ts:284]: same

File: pkg/apihandler/api.go (hunk 1 above)
Comment (jensneuse) [reply pkg/apihandler/api.go:18]: I'm not sure. I think anything is valid?

**Review by jensneuse — CHANGES_REQUESTED** [review]
A few minor issues.
One thing I was missing is that my understanding was that in dev mode (wunderctl up) we always want to set the same static value. Am I missing this?

File: pkg/apihandler/api.go (hunk 1 above)
Comment (Aenimus) [reply pkg/apihandler/api.go:18]: Let me rephrase. What would be reasonable to include?

File: pkg/apihandler/api.go [hunk 6]
```diff
@@ -25,6 +35,68 @@ type Options struct {
...
+	case WgEnvCsrfSecret:
+		cookieBasedSecrets.CsrfSecret = secretValue
+	case WgEnvBlockKey:
+		cookieBasedSecrets.BlockKey = secretValue
+	case WgEnvHashKey:
+		cookieBasedSecrets.HashKey = secretValue
+	}
+}
+
+func generateRandomStringOfLength(length int) []byte {
```
Comment (StarpTech) [inline pkg/apihandler/api.go:66]: `rand.Intn` should not be used for security-senstive stuff. 

> This package's outputs might be easily predictable regardless of how it's seeded. For random numbers suitable for security-sensitive work, see the crypto/rand package.
https://pkg.go.dev/math/rand

Use this

```go
// GenerateRandomBytes returns securely generated random bytes.
// It will return an error if the system's secure random
// number generator fails to function correctly, in which
// case the caller should not continue.
func GenerateRandomBytes(n int) ([]byte, error) {
	b := make([]byte, n)
	_, err := rand.Read(b)
	// Note that err == nil only if we read len(b) bytes.
	if err != nil {
		return nil, err
	}

	return b, nil
}

// GenerateRandomString generated a random string with the length as the first argument
func GenerateRandomString(n int) (string, error) {
	b, err := GenerateRandomBytes(n)
	return base64.URLEncoding.EncodeToString(b), err
}
```

File: packages/protobuf/wgpb/wundernode_config.ts [hunk 7]
```diff
@@ -771,9 +771,6 @@ export interface CookieBasedAuthentication {
   providers: AuthProvider[];
   authorizedRedirectUris: ConfigurationVariable[];
   authorizedRedirectUriRegexes: ConfigurationVariable[];
-  hashKey: ConfigurationVariable | undefined;
```
Comment (StarpTech) [inline packages/protobuf/wgpb/wundernode_config.ts:774]: Why do we remove this? IMO the user should still be able to configure it. The only difference is that we generate the value. This will also be a breaking change.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/apihandler/api.go [hunk 8]
```diff
@@ -25,6 +35,68 @@ type Options struct {
...
+	bytes := make([]byte, length)
+	for i := range bytes {
+		bytes[i] = validSecretCharacters[rand.Intn(len(validSecretCharacters))]
```
Comment (StarpTech) [inline pkg/apihandler/api.go:69]: `rand.Intn` should not be used for security-sensitive work. Please use crypto rand e.g.

```go
func GenerateRandomString(n int) (string, error) {
	const letters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-"
	ret := make([]byte, n)
	for i := 0; i < n; i++ {
		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(letters))))
		if err != nil {
			return "", err
		}
		ret[i] = letters[num.Int64()]
	}

	return string(ret), nil
}
```

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: packages/protobuf/wgpb/wundernode_config.ts (hunk 7 above)
Comment (Aenimus) [reply packages/protobuf/wgpb/wundernode_config.ts:774]: My mistake. I didn't intend to remove from interfaces.

Comment (Aenimus) [reply packages/protobuf/wgpb/wundernode_config.ts:774]: This file is generated from proto, and I removed the proto values, so that's why they're gone here.

Is this definitely a breaking change?

**Review by jensneuse — APPROVED** [review]

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: docs-website/src/pages/docs/self-hosted/security.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
+---
+title: 'Security'
+pageTitle: WunderGraph - Security
+description: Securing your self-hosted WunderGraph deployment
+---
+
+WunderGraph uses three secrets to secure your deployment.
```
Comment (StarpTech) [inline docs-website/src/pages/docs/self-hosted/security.md:7]: The secrets are only about authentication, nothing more.

File: docs-website/src/pages/docs/self-hosted/security.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+Cryptographically-secure (pseudo-)randomly generated strings should be added to your `.env` file:
+
+```js
+// these strings are insecure examples with the intention to display length
+WUNDERGRAPH_CSRF_TOKEN_SECRET = aaaaaaaaaaa; // 11 bytes
+WUNDERGRAPH_SECURE_COOKIE_HASH_KEY = aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa; // 32 bytes
+WUNDERGRAPH_SECURE_COOKIE_BLOCK_KEY = aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa; // 32 bytes
+```
+
+Although values will be generated for you in the case that they have not been supplied, these secrets are your
```
Comment (StarpTech) [inline docs-website/src/pages/docs/self-hosted/security.md:17]: We should mention that in case of you don't use fix values, we regenerate them and you users will be logged out after every restart

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/apihandler/api.go [hunk 11]
```diff
@@ -25,6 +36,80 @@ type Options struct {
...
+			return nil, err
+		}
+		bytes[i] = validSecretCharacters[num.Int64()]
+	}
+
+	return bytes, nil
+}
+
+func NewCookieBasedSecrets(isDevMode bool) (cookieBasedSecrets *CookieBasedSecrets, errorMessages []string) {
+	if isDevMode {
```
Comment (StarpTech) [inline pkg/apihandler/api.go:83]: I'd handle this condition outside of this method, so the only responsibility of this method is to generate "new" ones.

File: pkg/apihandler/apihandler.go [hunk 12]
```diff
@@ -2120,18 +2120,24 @@ func (r *Builder) registerAuth(insecureCookies bool) error {
 
 	if h := loadvariable.String(r.api.AuthenticationConfig.CookieBased.HashKey); h != "" {
 		hashKey = []byte(h)
+	} else if fallBackValue := r.api.CookieBasedSecrets.HashKey; fallBackValue != nil {
```
Comment (StarpTech) [inline pkg/apihandler/apihandler.go:2123]: I'll try :smile: in IT it is called `fallback`

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/validate/apiconfig_test.go [hunk 13]
```diff
@@ -44,142 +53,278 @@ func TestValidateApiConfig(t *testing.T) {
...
+				},
+			},
+			CookieBasedSecrets: &apihandler.CookieBasedSecrets{
+				CsrfSecret: elevenBytes,
+				BlockKey:   elevenBytes,
+			},
+		})
+
+		assert.False(t, valid)
+		assert.Equal(t, "the required environment variable WUNDERGRAPH_SECURE_COOKIE_BLOCK_KEY must have a length of 32", messages[0])
```
Comment (StarpTech) [inline pkg/validate/apiconfig_test.go:293]: Let's go with `WG_*` for consistency.

File: pkg/validate/apiconfig.go [hunk 14]
```diff
@@ -1,32 +1,51 @@
...
 
+// ApiConfig validates the Api
+func ApiConfig(api *apihandler.Api) (bool, []string) {
 	if !api.HasCookieAuthEnabled() {
 		// skip following tests if cookie auth is not enabled
 		// in this case, we don't have to check it
 		return true, nil
 	}
 
+	requiredCsrfLength := 11
```
Comment (StarpTech) [inline pkg/validate/apiconfig.go:22]: Where did you implement the logic for the cloud?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/validate/apiconfig.go (hunk 14 above)
Comment (StarpTech) [reply pkg/validate/apiconfig.go:22]: Clarified

**Review by StarpTech — APPROVED** [review]
LGTM
